### PR TITLE
Add kubeVersion checks for catalogv1

### PR DESCRIPTION
**Pull request by rmweir**
**Problem:**
Some charts will only be compatible with certain kubernetes versions

**Solution:**
Validate kubernetes version meets kubeVersion constraint in chart if one is set.

**Issue:**
https://github.com/rancher/rancher/issues/29346

### CODE REVIEW

File: pkg/api/norman/customization/multiclusterapp/store.go [hunk 1]
```diff
@@ -183,7 +185,22 @@ func (s *Store) validateRancherVersion(data map[string]interface{}) error {
...
 	}
 
-	return catUtil.ValidateRancherVersion(template)
+	if err := catUtil.ValidateRancherVersion(template); err != nil {
+		return err
+	}
+
+	targets, _ := data["targets"].([]interface{})
+	for _, target := range targets {
+		targetMap, _ := target.(map[string]interface{})
```
Comment (mrajashree) [inline pkg/api/norman/customization/multiclusterapp/store.go:194]:
```suggestion
		targetMap, ok := target.(map[string]interface{})
		if !ok {
		    continue
		}
```

File: pkg/catalog/utils/version.go [hunk 2]
```diff
@@ -63,6 +65,41 @@ func VersionGreaterThan(a, b string) bool {
...
+	k8sVersion, err := semver.Parse(strings.TrimPrefix(cluster.Status.Version.String(), "v"))
+	if err != nil {
+		return err
+	}
+	if !constraint(k8sVersion) {
+		return httperror.NewAPIError(httperror.InvalidBodyContent, fmt.Sprintf("incompatible kubernetes version [%s] for template [%s]", k8sVersion.String(), template.Name))
+	}
+	return nil
+}
+
```
Comment (cbron) [inline pkg/catalog/utils/version.go:102]: Trying to wrap my head around the flows going on here, can you summarize which methods are primary ones and how they are being used. Somewhat confusing as there are several with the same name.

Comment (rmweir) [reply pkg/catalog/utils/version.go:102]: This method, ValidateKubeVersion, in the utils package is the main one being added. Before, most charts were just being checked for their compatibility with the Rancher server version with a method called ValidateRancherVersion. 

Now both of these methods, ValidateKubeVersion and ValidateRancherVersion are in a new method called ValidateChartCompatibility. Most occurrences of ValidateRancherVersion have been replaced with this function. Any other functions called validateChartCompatibility are helper functions in their own respective packages that prepare their data for the functions and do any additional processing. This is because compatibility is validated in stores, validators, controllers, and action handlers which may all vary a bit in how they extract the information needed.

File: pkg/api/norman/store/app/store.go [hunk 3]
```diff
@@ -57,11 +64,16 @@ func (s *Store) Update(apiContext *types.APIContext, schema *types.Schema, data
 		return nil, err
 	}
 
-	if err := s.validateRancherVersion(data); err != nil {
+	if err := s.validateForMultiClusterApp(id, "update"); err != nil {
 		return nil, err
 	}
 
-	if err := s.validateForMultiClusterApp(id, "update"); err != nil {
+	clusterName, err := s.parseClusterID(apiContext.Request.URL.Path)
```
Comment (brendarearden) [inline pkg/api/norman/store/app/store.go:71]: nit: var name is clusterName, but method is parseClusterID. not super familiar with this area, so my bad if those are commonly interchangeable.

**Review by brendarearden — APPROVED** [review]
LGTM

File: pkg/catalog/utils/version.go [hunk 4]
```diff
@@ -63,6 +65,41 @@ func VersionGreaterThan(a, b string) bool {
...
+func ValidateKubeVersion(template *v3.CatalogTemplateVersion, clusterlister v3.ClusterLister, clusterName string) error {
+	if template.Spec.KubeVersion == "" {
+		return nil
+	}
+	constraint, err := semver.ParseRange(template.Spec.KubeVersion)
+	if err != nil {
+		logrus.Errorf("failed to parse constraint for kubeversion %s: %v", template.Spec.KubeVersion, err)
+		return nil
```
Comment (mrajashree) [inline pkg/catalog/utils/version.go:85]: q: why don't we return the error here?

File: pkg/catalog/utils/version.go [hunk 5]
```diff
@@ -63,6 +65,41 @@ func VersionGreaterThan(a, b string) bool {
...
+	}
+
+	cluster, err := clusterlister.Get("", clusterName)
```
Comment (mrajashree) [inline pkg/catalog/utils/version.go:88]: nit: `clusterLister` instead of `clusterlister`

File: pkg/catalog/utils/version.go (hunk 4 above)
Comment (rmweir) [reply pkg/catalog/utils/version.go:85]: Because an error in this function insinuates that the cluster version is not compatible with the chart, but if the KubeVersion constraint is not readable I think it's best to just treat it as though it's empty and install the app. Installing an app that isn't compatible won't be catastrophic.

**Review by mrajashree — DISMISSED** [review]

File: pkg/api/norman/store/app/store.go (hunk 3 above)
Comment (rmweir) [reply pkg/api/norman/store/app/store.go]: They are but I change it anyway

**Review by mrajashree — APPROVED** [review]
lgtm

**Review by brendarearden — APPROVED** [review]

File: pkg/api/norman/store/app/store.go [hunk 6]
```diff
@@ -175,3 +187,20 @@ func (s *Store) parseAppExternalID(data map[string]interface{}) (string, string,
...
+	}
+	if _, err := s.ClusterLister.Get("", parts[0]); err != nil {
+		if errors.IsNotFound(err) {
+			return "", httperror.NewAPIError(httperror.NotFound, fmt.Sprintf("cannot find target cluster [%s]: %v", parts[0], err))
+		}
+		return "", httperror.NewAPIError(httperror.ServerError, fmt.Sprintf("error getting target custer [%s]: %v", parts[0], err))
+
+	}
+	return parts[0], nil
+}
```
Comment (cbron) [inline pkg/api/norman/store/app/store.go:206]: Can we not get project name at all from apiContext and then use `ref.parse` ? There is also a `GetClusterID` method that might work. Adding all these things feels dangerous.

File: pkg/catalog/utils/version.go [hunk 7]
```diff
@@ -95,7 +132,7 @@ func ReleaseServerVersion(serverVersion string) bool {
 	return true
 }
 
-func LatestAvailableTemplateVersion(template *v3.CatalogTemplate) (*v32.TemplateVersionSpec, error) {
+func LatestAvailableTemplateVersion(template *v3.CatalogTemplate, clusterLister v3.ClusterLister, clusterName string) (*v32.TemplateVersionSpec, error) {
```
Comment (cbron) [inline pkg/catalog/utils/version.go:135]: This `LatestAvailableTemplateVersion` subs as a `ValidateChartCompatibility` parent ? We should at least clarify that with a comment above.

**Review by cbron — COMMENTED** [review]
Some minor things. In general this approach feels a bit jumbled, we are now passing around a lot of objects throughout a lot of method signatures just to eventually hand them off to a util method down the line that is mostly unrelated to the functionality of the feature. But I don't have a better idea right now.

File: pkg/api/norman/store/app/store.go (hunk 6 above)
Comment (rmweir) [reply pkg/api/norman/store/app/store.go:206]: it's because this is a project scoped api

File: pkg/catalog/utils/version.go (hunk 7 above)
Comment (rmweir) [reply pkg/catalog/utils/version.go:135]: This is technically finds the latest available compatible template version, it did the same prior but only checked Rancher version constraints.

File: pkg/api/norman/store/app/store.go (hunk 6 above)
Comment (dramich) [reply pkg/api/norman/store/app/store.go:206]: @rmweir I don't understand your answer, why can't `GetClusterID` be used here?

**Review by dramich — CHANGES_REQUESTED** [review]
There are a couple of options we can do with this:
1. Move version utils and let them live on the catalog manager struct, then add that to the scaled/management context, similar to what we do with `UserManager`, then wherever things are being created and registered grab the catalogManager struct. This will save a lot of work in the future if we ever need to extend what this is doing and need access to more controllers/listers. 
2. "Register" the version piece on its own by creating a var at the top which holds the controllers/listers and all the currently exposed functions could then use

I think option 1 is better and follows patterns we already have in place.

File: pkg/api/norman/store/app/store.go (hunk 6 above)
Comment (rmweir) [reply pkg/api/norman/store/app/store.go:206]: @dramich that function is intended to be used on cluster requests. It does not work because the url in this case is of the form: `/v3/project(s)/<clusterName>:<projectName>/app`, but it is expecting `/v3|k8s/cluster(s)/<clusterName>`. I could modify GetClusterID so it will work.

File: pkg/types/config/context.go [hunk 8]
```diff
@@ -72,6 +75,14 @@ type ScaledContext struct {
...
 }
 
+type CatalogManager interface {
+	ValidateChartCompatibility(template *managementv3.CatalogTemplateVersion, clusterName string) error
+	ValidateKubeVersion(template *managementv3.CatalogTemplateVersion, clusterName string) error
+	ValidateRancherVersion(template *managementv3.CatalogTemplateVersion) error
+	LatestAvailableTemplateVersion(template *managementv3.CatalogTemplate, clusterName string) (*v32.TemplateVersionSpec, error)
+	GetSystemAppCatalogID(templateVersionID, clusterName string) (string, error)
+}
+
```
Comment (dramich) [inline pkg/types/config/context.go:85]: This should be in its own package or under catalog, it doesn't belong in context

File: pkg/catalog/manager/manager.go [hunk 9]
```diff
@@ -2,16 +2,24 @@ package manager
...
+	"sort"
 	"strings"
 	"time"
 
+	"github.com/blang/semver"
+	mVersion "github.com/mcuadros/go-version"
+	"github.com/rancher/norman/httperror"
+	v32 "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
+	"github.com/rancher/rancher/pkg/catalog/utils"
+
```
Comment (dramich) [inline pkg/catalog/manager/manager.go:14]: extra space here

File: pkg/catalog/manager/manager.go [hunk 10]
```diff
@@ -184,3 +194,108 @@ func (m *Manager) deleteBadCatalogTemplates() []error {
...
+	if err != nil {
+		return err
+	}
+
+	k8sVersion, err := semver.Parse(strings.TrimPrefix(cluster.Status.Version.String(), "v"))
+	if err != nil {
+		return err
+	}
+	if !constraint(k8sVersion) {
+		return httperror.NewAPIError(httperror.InvalidBodyContent, fmt.Sprintf("incompatible kubernetes version [%s] for template [%s]", k8sVersion.String(), template.Name))
```
Comment (dramich) [inline pkg/catalog/manager/manager.go:228]: None of the functions in here should return an APIerror, these helpers are getting called from all over and the caller should decide if they need to wrap as an APIerror

**Review by dramich — DISMISSED** [review]

**Review by cbron — APPROVED** [review]

File: pkg/catalog/manager/manager_test.go [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+					{
+						ExternalID: "catalog://?catalog=library&template=artifactory-ha&version=0.12.14",
+						Version:    "0.12.14",
+					},
+				},
+			},
+		},
+	}
+
+	testLatestAvailableTemplateVersion(t, "v2.1.0", "0.12.15", template)
```
Comment (dramich) [inline pkg/catalog/manager/manager_test.go:60]: Build up a map of your test cases then in your test func loop over the map and process the cases
In the test func you can create one assert at the top with assert.New(t)

Comment (dramich) [reply pkg/catalog/manager/manager_test.go:60]: Oh this already existed...

**Review by dramich — APPROVED** [review]
